File: maze_solver.py
class Square:
    WALL = "#"
    PATH = " "

Point = tuple[int, int]
Path = list[Point]

Map = list[list[str]]
SeenMap = list[list[bool]]

directions = (
    (-1, 0),
    (0, -1),
    (1, 0),
    (0, 1),
)

def walk(maze: Map, current: Point, seen: SeenMap, end: Point, path: Path) -> bool:
    # off grid
    if (current[0] < 0 or current[0] > len(maze[0])) or (current[1] < 0 or current[1] > len(maze)):
        return False

    # already visited square
    if seen[current[1]][current[0]] is True:
        return False
    
    # ran into a wall
    if maze[current[1]][current[0]] == Square.WALL:
        return False

    # found end
    if current[0] == end[0] and current[1] == end[1]:
        path.append(current)
        return True

    # pre - push to stack
    seen[current[1]][current[0]] = True
    path.append(current)

    # recurse
    for direction in directions:
        next_point = (current[0] + direction[0], current[1] + direction[1])

        if walk(maze, next_point, seen, end, path):
            return True

    # post - pop from stack
    path.pop()

    # dead-end - return to the previous step
    return False

def solve(maze: Map, start: Point, end: Point) -> Path:
    path = []
    seen = [[False] * len(maze[0]) for _ in range(len(maze))]

    walk(maze, start, seen, end, path)

    return path

if __name__ == "__main__":
    maze = [
        ["#", "#", "#", "#", "#", "E", "#",],
        ["#", " ", " ", " ", " ", " ", "#",],
        ["#", "S", "#", "#", "#", "#", "#",],
    ]
    start = (1, 2)
    end = (5, 0)

    path = solve(maze, start, end)
    print(f"{start=}")
    print(f"{end=}")
    print(f"{path=}")

File: maze-solver/maze_solver.go
package maze_solver

import (
  "errors"
)

type Point struct {
  x, y int
}

var directions = [4]Point{
  {-1, 0},
  {0, -1},
  {1, 0},
  {0, 1},
}

type Maze [][]rune;

func Solve(maze Maze, start Point, end Point) ([]Point, error) {
  stack := make([]Point, 0)
  visited := make(map[Point]bool)

  if _solve(maze, end, start, &stack, visited) {
    return stack, nil
  }

  return make([]Point, 0, 0), errors.New("no path found")
}

func _solve(maze Maze, end Point, current Point, stack *[]Point, visited map[Point]bool) bool {
  if (current.y >= len(maze) || current.y < 0) || (current.x >= len(maze[0]) || current.x < 0) {
    return false
  }

  if visited[current] {
    return false
  }

  if maze[current.y][current.x] == '#' {
    return false
  }

  if current == end {
    *stack = append(*stack, current)
    return true
  }

  visited[current] = true
  *stack = append(*stack, current)

  for _, d := range directions {
    next := Point{x: current.x + d.x, y: current.y + d.y}

    if _solve(maze, end, next, stack, visited) {
      return true
    }
  }

  *stack = (*stack)[:len(*stack)-1]
  return false
}
